lib/IDE/CompletionInstance.cpp

```diff
@@ -162,10 +165,77 @@ static DeclContext *getEquivalentDeclContextFromSourceFile(DeclContext *DC,
   return newDC;
 }
 
+/// Check if any dependent files are modified since \p timestamp.
+bool areAnyDependentFilesInvalidated(CompilerInstance &CI,
+                                     llvm::vfs::FileSystem &FS,
+                                     StringRef currentFileName,
+                                     llvm::sys::TimePoint<> timestamp) {
+
+  auto isInvalidated = [&](StringRef filePath) -> bool {
+    auto stat = FS.status(filePath);
+    if (!stat)
+      // Missing.
+      return true;
+
+    auto lastModTime = stat->getLastModificationTime();
+    if (lastModTime > timestamp)
+      // Modified.
+      return true;
+
+    // If the last modification time is zero, this file is probably from a
+    // virtual file system. We need to check the content.
+    if (lastModTime == llvm::sys::TimePoint<>()) {
+      if (&CI.getFileSystem() == &FS)
+        return false;
+
+      auto oldContent = CI.getFileSystem().getBufferForFile(filePath);
+      auto newContent = FS.getBufferForFile(filePath);
+      if (!oldContent || !newContent)
+        // (unreachable?)
+        return true;
+
+      if (oldContent.get()->getBuffer() != newContent.get()->getBuffer())
+        // Different content.
+        return true;
+    }
+
+    return false;
+  };
+
+  // Check files in the current module.
+  for (FileUnit *file : CI.getMainModule()->getFiles()) {
+    StringRef filename;
+    if (auto SF = dyn_cast<SourceFile>(file))
+      filename = SF->getFilename();
+    else if (auto LF = dyn_cast<LoadedFile>(file))
+      filename = LF->getFilename();
+    else
+      continue;
+
+    // Ignore the current file and synthesized files.
+    if (filename.empty() || filename.front() == '<' ||
+        filename.equals(currentFileName))
```

Comment:
Comparing filename seems potentially brittle - shouldn't we know which file is the completion buffer directly without looking at the name?

Reply:
Updated the PR.